akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFoldAsyncSpec.scala

```diff
@@ -0,0 +1,63 @@
+/**
+ * Copyright (C) 2014-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.scaladsl
+
+import akka.NotUsed
+
+import scala.concurrent.{ Await, Future }
+import scala.util.control.NoStackTrace
+
+import akka.stream.ActorMaterializer
+import akka.testkit.AkkaSpec
+import akka.stream.testkit.Utils._
+import scala.concurrent.duration._
+
+class FlowFoldAsyncSpec extends AkkaSpec {
+  implicit val materializer = ActorMaterializer()
+  implicit def ec = materializer.executionContext
+
+  "A FoldAsync" must {
+    val input = 1 to 100
+    val expected = input.sum
+    val inputSource = Source(input).filter(_ ⇒ true).map(identity)
+    val foldSource = inputSource.foldAsync[Int](0) { (a, b) ⇒
+      Future(a + b)
+    }.filter(_ ⇒ true).map(identity)
+    val foldFlow = Flow[Int].filter(_ ⇒ true).map(identity).foldAsync(0) {
+      (a, b) ⇒ Future { Thread.sleep(100); a + b }
+    }.filter(_ ⇒ true).map(identity)
+    val foldSink = Sink.foldAsync[Int, Int](0) { (a, b) ⇒ Future(a + b) }
+
+    "work when using Source.foldAsync" in assertAllStagesStopped {
+      Await.result(foldSource runWith Sink.head, 3.seconds) should be(expected)
+    }
+
+    "work when using Sink.foldAsync" in assertAllStagesStopped {
+      Await.result(inputSource runWith foldSink, 3.seconds) should be(expected)
+    }
+
+    "work when using Flow.foldAsync" in assertAllStagesStopped {
+      Await.result(inputSource via foldFlow runWith Sink.head, 12.seconds) should be(expected)
+    }
+
+    "work when using Source.foldAsync + Flow.foldAsync + Sink.foldAsync" in assertAllStagesStopped {
+      Await.result(foldSource via foldFlow runWith foldSink, 3.seconds) should be(expected)
+    }
+
+    "propagate an error" in assertAllStagesStopped {
+      val error = new Exception with NoStackTrace
+      val future = inputSource.map(x ⇒ if (x > 50) throw error else x).runFoldAsync[NotUsed](NotUsed)(Keep.noneAsync)
+      the[Exception] thrownBy Await.result(future, 3.seconds) should be(error)
+    }
+
+    "complete future with failure when folding function throws" in assertAllStagesStopped {
```

Comment:
@johanandren I'm trying to adapt tests using `TestLatch`, but I guess there is something I don't get about it, as it timeout after the first element.

```scala
    "signal future failure" in assertAllStagesStopped {
      val latch = TestLatch(5)
      val c = TestSubscriber.manualProbe[Unit]()
      implicit val ec = system.dispatcher
      val p = Source(1 to 5).foldAsync({}) { (_, n) ⇒
        Future {
          if (n == 3) throw new RuntimeException("err1") with NoStackTrace
          else {
            Await.ready(latch, 500.milliseconds) // TODO: 10s
            ()
          }
        }
      }.to(Sink.fromSubscriber(c)).run()

      val sub = c.expectSubscription()
      sub.request(10)
      c.expectError().getMessage should be("err1")
      latch.countDown()
    }
```